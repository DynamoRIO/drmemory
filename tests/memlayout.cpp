/* **********************************************************
 * Copyright (c) 2020 Google, Inc.  All rights reserved.
 * **********************************************************/

/* Dr. Memory: the memory debugger
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation;
 * version 2.1 of the License, and no later version.

 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Library General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "drmemory_annotations.h"
#include <iostream>

int main()
{
    int i,**j,k,l,*m;
    i = 0;
    j = new int*[3];
    j[0] = new int;
    j[1] = &i;
    m = *(j+1);
    DRMEMORY_ANNOTATE_DUMP_MEMORY_LAYOUT();
    j[1] = &k;
    k=10;
    *(j[0]) = 5;
    j[2] = j[0];
    *(j[0]) = 18;
    *m = 4;
    l = 3;

    char *ch = new char[13];
    ch[4] = 'x';

    DRMEMORY_ANNOTATE_DUMP_MEMORY_LAYOUT();

    std::cerr << "goodbye\n";

    return 0;
}
