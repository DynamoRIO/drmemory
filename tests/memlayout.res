# **********************************************************
# Copyright (c) 2020 Google, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
{
  "heap objects": [
    {
      "address": "",
%if X32
      "size": "12",
%endif
%if X64
      "size": "24",
%endif
      "contents": [
        {
          "address": "",
          "value": "",
          "points-to-type": "heap",
          "points-to-base": "",
          "points-to-offset": ""
        },
# We just make sure we have a stack section.
  "thread stacks": [
