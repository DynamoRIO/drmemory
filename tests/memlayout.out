# **********************************************************
# Copyright (c) 2020 Google, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
~~Dr.M~~ Memory layout written to:
~~Dr.M~~ Memory layout written to:
goodbye
~~Dr.M~~ ERRORS FOUND:
~~Dr.M~~       0 unique,     0 total unaddressable access(es)
~~Dr.M~~       0 unique,     0 total uninitialized access(es)
~~Dr.M~~       0 unique,     0 total invalid heap argument(s)
~~Dr.M~~       0 unique,     0 total warning(s)
%if X32
~~Dr.M~~       2 unique,     2 total,     29 byte(s) of leak(s)
%endif
%if X64
~~Dr.M~~       2 unique,     2 total,     41 byte(s) of leak(s)
%endif
~~Dr.M~~       0 unique,     0 total,      0 byte(s) of possible leak(s)
